### PR TITLE
Moving rules to a separate folder

### PR DESCRIPTION
**Pull request by dczerwinski**
#### What has been done to verify that this works as intended?
`checkTests` runs successfully.

#### Why is this the best possible solution? Were any other approaches considered?
All test rules have been grouped and moved to one folder.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
It should not affect users.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [ ] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [ ] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [ ] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CallbackCountingTaskExecutorRule.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package org.odk.collect.android.support.rules
+
+import androidx.arch.core.executor.testing.CountingTaskExecutorRule
+
+class CallbackCountingTaskExecutorRule : CountingTaskExecutorRule() {
+
+    private var finishedCallback: Runnable? = null
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CallbackCountingTaskExecutorRule.kt:7]: There is no need to make it nullable I think.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.odk.collect.android.support.pages.ShortcutsPage
+import java.util.function.Consumer
+
+class CollectTestRule @JvmOverloads constructor(
+    private val useDemoProject: Boolean = true
+) : TestRule {
+
+    override fun apply(base: Statement, description: Description): Statement {
+        return object : Statement() {
+            @Throws(Throwable::class)
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt]: Let's get rid of it if it throws something it throws it and we do not catch it anywhere so I don't like those annotations in rules.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        val scenario = ActivityScenario.launch(AndroidShortcutsActivity::class.java)
+        return ShortcutsPage(scenario).assertOnPage()
+    }
+
+    fun <T : Page<T>?> launch(intent: Intent, destination: T): T {
+        ActivityScenario.launch<Activity>(intent)
+        return destination!!.assertOnPage()
+    }
+
+    fun <T : Page<T>?> launchForResult(
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt]: This shouldn't be nullable I think.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            .clickManuallyEnterProjectDetails()
+            .inputUrl(serverUrl!!)
+            .addProject()
+
+    fun launchShortcuts(): ShortcutsPage {
+        val scenario = ActivityScenario.launch(AndroidShortcutsActivity::class.java)
+        return ShortcutsPage(scenario).assertOnPage()
+    }
+
+    fun <T : Page<T>?> launch(intent: Intent, destination: T): T {
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt]: This shouldn't be nullable I think.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                base.evaluate()
+            }
+        }
+    }
+
+    fun startAtMainMenu() = MainMenuPage()
+
+    fun startAtFirstLaunch() = FirstLaunchPage()
+
+    fun withProject(serverUrl: String?): MainMenuPage =
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt]: This shouldn't be nullable I think.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/IdlingResourceRule.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    override fun apply(base: Statement, description: Description): Statement =
+        IdlingResourceStatement(idlingResources, base)
+
+    private class IdlingResourceStatement(
+        private val idlingResources: List<IdlingResource>,
+        private val base: Statement
+    ) : Statement() {
+
+        @Throws(Throwable::class)
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/rules/IdlingResourceRule.kt]: Let's remove it.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/RunnableRule.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import org.junit.runner.Description
+import org.junit.runners.model.Statement
+
+class RunnableRule(
+    private val runnable: Runnable
+) : TestRule {
+
+    override fun apply(base: Statement, description: Description): Statement =
+        object : Statement() {
+            @Throws(Throwable::class)
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/rules/RunnableRule.kt]: Let's remove it.

File: collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import java.io.IOException
+import java.util.ArrayList
+
+class MockWebServerRule : TestRule {
+
+    private val mockWebServers: MutableList<MockWebServer> = ArrayList()
+
+    override fun apply(base: Statement, description: Description): Statement =
+        object : Statement() {
+            @Throws(Throwable::class)
```
Comment (grzesiek2010) [inline collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt]: Let's remove it.

File: collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            override fun evaluate() {
+                try {
+                    base.evaluate()
+                } finally {
+                    teardown()
+                }
+            }
+        }
+
+    @Throws(Exception::class)
```
Comment (grzesiek2010) [inline collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt]: Let's remove it.

File: collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    fun start(): MockWebServer {
+        val mockWebServer = MockWebServer()
+        mockWebServers.add(mockWebServer)
+
+        mockWebServer.start()
+        return mockWebServer
+    }
+
+    @Throws(IOException::class)
```
Comment (grzesiek2010) [inline collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt]: Let's remove it.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt (hunk 10 above)
Comment (dczerwinski) [reply collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt]: Done.

File: collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt (hunk 9 above)
Comment (dczerwinski) [reply collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt]: Done.

File: collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt (hunk 8 above)
Comment (dczerwinski) [reply collect_app/src/test/java/org/odk/collect/android/rules/MockWebServerRule.kt]: Done.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/RunnableRule.kt (hunk 7 above)
Comment (dczerwinski) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/RunnableRule.kt]: Done.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/IdlingResourceRule.kt (hunk 6 above)
Comment (dczerwinski) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/IdlingResourceRule.kt]: Done.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt (hunk 5 above)
Comment (dczerwinski) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt]: Done.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt (hunk 4 above)
Comment (dczerwinski) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt]: Done.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt (hunk 3 above)
Comment (dczerwinski) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt]: Done.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt (hunk 2 above)
Comment (dczerwinski) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CollectTestRule.kt]: Done.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CallbackCountingTaskExecutorRule.kt (hunk 1 above)
Comment (dczerwinski) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CallbackCountingTaskExecutorRule.kt:7]: As I checked it probably need to be nullable. If I change it to not nullable then some tests are failing.

Comment (grzesiek2010) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CallbackCountingTaskExecutorRule.kt:7]: Which test for example?

Comment (dczerwinski) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CallbackCountingTaskExecutorRule.kt:7]: All tests in ConfigureWithQRCodeTest class.

Comment (grzesiek2010) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/rules/CallbackCountingTaskExecutorRule.kt:7]: Ok now I see.

**Comment by grzesiek2010** [comment]
@dczerwinski please fix conflicts and also make sure checkboxes in your pr's description are selected (of course if you verified cases they represent).

**Review by grzesiek2010 — APPROVED** [review]

**Pull request merged**